session.go

```diff
@@ -4240,12 +4248,20 @@ func (q *Query) Apply(change Change, result interface{}) (info *ChangeInfo, err
 	session.SetMode(Strong, false)
 
 	var doc valueResult
-	err = session.DB(dbname).Run(&cmd, &doc)
-	if err != nil {
-		if qerr, ok := err.(*QueryError); ok && qerr.Message == "No matching object found" {
+	for {
+		err = session.DB(dbname).Run(&cmd, &doc)
+
+		if err == nil {
+			break
+		} else if change.Upsert && IsDup(err) {
+			// Retry duplicate key errors on upserts.
+			// https://docs.mongodb.com/v3.2/reference/method/db.collection.update/#use-unique-indexes
+			continue
+		} else if qerr, ok := err.(*QueryError); ok && qerr.Message == "No matching object found" {
 			return nil, ErrNotFound
+		} else {
+			return nil, err
```

Comment:
Please take this one out of the else and after the if blocks. Makes it more clear that this isn't looping around no matter what.